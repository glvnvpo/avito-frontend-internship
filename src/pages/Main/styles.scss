@import '~antd/dist/antd.css';
@import "../../style/variables.scss";

.main {
  display: flex;
  justify-content: center;

  .content {
    width: 50rem;
    padding-right: 10px;
    padding-left: 10px;

    .header {
      border-bottom: 2px solid $grey;
      display: flex;
      justify-content: space-between;
      align-items: center;
      padding-top: 20px;
      padding-bottom: 20px;
      position: sticky;
      top: 0;

      .title {
        font-size: 1.5rem;
      }

      @media screen and (max-width: $medium) {
        padding-top: 10px;
        padding-bottom: 10px;

        .title {
          font-size: 1.25rem;
        }
      }
    }

    .cards {
      display: flex;
      flex-direction: column;
      align-items: center;
      margin-bottom: 50px;

      .empty {
        font-size: 1rem;
      }
    }
  }
}
