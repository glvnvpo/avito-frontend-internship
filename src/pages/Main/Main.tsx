// @flow

import React, {FC, useEffect, useState, MouseEvent} from 'react';
import {useLocation} from 'react-router-dom';
import {Button} from 'antd';
import axios from 'axios';
import {isEmpty, isNull} from 'lodash';
import './styles.scss';
import {ID, Story} from '../../types';
import {StoryCard} from '../../components/StoryCard';
import {Spinner} from '../../components/Spinner';
import {setStories} from '../../store/stories';
import {MINUTE} from '../../constants/time';
import {ITEM, NEW_STORIES} from '../../api/constants';
import {useAppDispatch, useAppSelector} from '../../hooks/redux';
import {MAIN_PAGE_PATH} from '../../routing/constants';

const STORIES_NUMBER = 100;

export const Main: FC = () => {

	let stories = useAppSelector(state => state.stories.value);

	const dispatch = useAppDispatch();
	
	const [isLoading, setLoading] = useState<boolean>(stories.length <= 0);
	
	let location = useLocation();
	
	let timer: NodeJS.Timeout | undefined;

	useEffect(() => {
		loadStoriesEachMinute();
		return () => clearTimeout(timer);
	}, [location?.pathname]);

	const loadStoriesEachMinute = () => {
		loadNewStories();

		timer = setTimeout(loadStoriesEachMinute, MINUTE);
	};

	const loadNewStories = (event?: MouseEvent) => {

		if (event) {
			setLoading(true);
		}

		axios<ID[]>(NEW_STORIES())
			.then(async ({data}) => {
				const newestStories = [...data.slice(0, STORIES_NUMBER)];
				let loadedStories: Story[] = [];
				const promises = newestStories.map(loadOneStory);

				await Promise.allSettled(promises)
					.then(data => 
						loadedStories = data
							.filter(({status}) => status === 'fulfilled')
							// @ts-ignore
							.map(({value}) => value)
					)
					.finally(() => {
						dispatch(setStories(loadedStories));
						setLoading(false);
					});
			})
			.catch(() => setLoading(false));
	};

	const loadOneStory = (id: ID) => {
		return new Promise((resolve, reject) => {
			axios<Story>(ITEM(id))
				.then(({data}) => !isNull(data) ? resolve(data) : reject())
				.catch(err => reject(err));
		});
	};

	return (
		<div className='main'>
			<div className='content'>
				<div className='header bg-white'>
					<span className='title bold color-grey2'>Latest news</span>
					<Button onClick={(e) => loadNewStories(e)} type='primary'>Update news</Button>
				</div>

				<div className='cards mt-10'>
					{ isLoading ? <Spinner className='mt-20'/> :
						(!isEmpty(stories)) ? stories.map((story: Story) =>
							<StoryCard
								story={story}
								key={story.id}
								asLink
								to={`${MAIN_PAGE_PATH}/${story.id}`}
							/>
						) : <span className='empty mt-20 bold color-dark-grey'>No stories found :(</span>
					}
				</div>
			</div>
		</div>
	);
};