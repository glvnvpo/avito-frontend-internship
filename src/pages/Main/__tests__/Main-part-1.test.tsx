import React from 'react';
import {mount, ReactWrapper} from 'enzyme';
import axios from 'axios';
import {Provider} from 'react-redux';
import {BrowserRouter} from 'react-router-dom';
import {createStore} from '@reduxjs/toolkit';
import {Main} from '../Main';
import {NEW_STORIES as mockNEW_STORIES} from '../../../api/constants';
import {waitForComponentToPaint} from '../../../jest/helpers/wait-for-component-to-paint';

// Тесты для страницы Main разделены на несколько файлов т.к. act()
// падает с ошибкой при вызове несколько раз в одном файле

jest.mock('axios', () => ({
	__esModule: true,
	...jest.requireActual('axios'),
	default: jest.fn((url) => {
		if (url === mockNEW_STORIES()) {
			return Promise.resolve({data: [10]});
		} else return Promise.resolve({
			data: {
				id: 10,
				title: 'someTitle',
				by: 'coolUser',
				time: 1666689031,
				score: 3,
				url: 'https://reactjs.org/',
				descendants: 10
			}
		});
	})
}));

describe('Main Part 1', () => {

	let wrapper: ReactWrapper<any, React.Component['state'], React.Component> | undefined;

	let initValue = [{
		id: 20,
		title: 'someTitle',
		by: 'coolUser',
		time: 1666689031,
		score: 4,
		url: 'https://reactjs.org/',
		descendants: 10
	}];
	
	const getWrapper = (initStore = initValue) => {
		const store = createStore(() => ({
			stories: {
				value: initStore
			}
		}));

		return mount(<Provider store={store}>
			<BrowserRouter>
				<Main/>
			</BrowserRouter>
		</Provider>);
	};

	beforeEach(() => {
		jest.clearAllMocks();
	});
	
	it('should render Main page with StoryCard', () => {
		wrapper = getWrapper();
		expect(wrapper).toMatchSnapshot();
	});
	
	it('should render Main page with Spinner', () => {
		wrapper = getWrapper([]);
		waitForComponentToPaint();
		expect(wrapper).toMatchSnapshot();
	});
});