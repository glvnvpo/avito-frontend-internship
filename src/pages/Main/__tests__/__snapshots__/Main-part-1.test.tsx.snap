// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Part 1 should render Main page with Spinner 1`] = `
<Provider
  store={
    {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      location={
        {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        }
      }
      navigationType="POP"
      navigator={
        {
          "action": "POP",
          "createHref": [Function],
          "go": [Function],
          "listen": [Function],
          "location": {
            "hash": "",
            "key": "default",
            "pathname": "/",
            "search": "",
            "state": null,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Main>
        <div
          className="main"
        >
          <div
            className="content"
          >
            <div
              className="header bg-white"
            >
              <span
                className="title bold color-grey2"
              >
                Latest news
              </span>
              <Button
                onClick={[Function]}
                type="primary"
              >
                <ForwardRef
                  disabled={false}
                >
                  <InternalWave
                    disabled={false}
                  >
                    <button
                      className="ant-btn ant-btn-primary"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      <LoadingIcon
                        existIcon={false}
                        loading={false}
                        prefixCls="ant-btn"
                      >
                        <CSSMotion
                          motionName="ant-btn-loading-icon-motion"
                          onAppearActive={[Function]}
                          onAppearStart={[Function]}
                          onEnterActive={[Function]}
                          onEnterStart={[Function]}
                          onLeaveActive={[Function]}
                          onLeaveStart={[Function]}
                          removeOnLeave={true}
                          visible={false}
                        >
                          <DomWrapper />
                        </CSSMotion>
                      </LoadingIcon>
                      <span
                        key=".0"
                      >
                        Update news
                      </span>
                    </button>
                  </InternalWave>
                </ForwardRef>
              </Button>
            </div>
            <div
              className="cards mt-10"
            >
              <Spinner
                className="mt-20"
              >
                <div
                  className="spinner orange"
                >
                  <Spin
                    className="mt-20"
                    size="large"
                  >
                    <Spin
                      className="mt-20"
                      size="large"
                      spinPrefixCls="ant-spin"
                    >
                      <div
                        aria-busy={true}
                        aria-live="polite"
                        className="ant-spin ant-spin-lg ant-spin-spinning mt-20"
                      >
                        <span
                          className="ant-spin-dot ant-spin-dot-spin"
                        >
                          <i
                            className="ant-spin-dot-item"
                          />
                          <i
                            className="ant-spin-dot-item"
                          />
                          <i
                            className="ant-spin-dot-item"
                          />
                          <i
                            className="ant-spin-dot-item"
                          />
                        </span>
                      </div>
                    </Spin>
                  </Spin>
                </div>
              </Spinner>
            </div>
          </div>
        </div>
      </Main>
    </Router>
  </BrowserRouter>
</Provider>
`;

exports[`Main Part 1 should render Main page with StoryCard 1`] = `
<Provider
  store={
    {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      location={
        {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        }
      }
      navigationType="POP"
      navigator={
        {
          "action": "POP",
          "createHref": [Function],
          "go": [Function],
          "listen": [Function],
          "location": {
            "hash": "",
            "key": "default",
            "pathname": "/",
            "search": "",
            "state": null,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Main>
        <div
          className="main"
        >
          <div
            className="content"
          >
            <div
              className="header bg-white"
            >
              <span
                className="title bold color-grey2"
              >
                Latest news
              </span>
              <Button
                onClick={[Function]}
                type="primary"
              >
                <ForwardRef
                  disabled={false}
                >
                  <InternalWave
                    disabled={false}
                  >
                    <button
                      className="ant-btn ant-btn-primary"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      <LoadingIcon
                        existIcon={false}
                        loading={false}
                        prefixCls="ant-btn"
                      >
                        <CSSMotion
                          motionName="ant-btn-loading-icon-motion"
                          onAppearActive={[Function]}
                          onAppearStart={[Function]}
                          onEnterActive={[Function]}
                          onEnterStart={[Function]}
                          onLeaveActive={[Function]}
                          onLeaveStart={[Function]}
                          removeOnLeave={true}
                          visible={false}
                        >
                          <DomWrapper />
                        </CSSMotion>
                      </LoadingIcon>
                      <span
                        key=".0"
                      >
                        Update news
                      </span>
                    </button>
                  </InternalWave>
                </ForwardRef>
              </Button>
            </div>
            <div
              className="cards mt-10"
            >
              <StoryCard
                asLink={true}
                extraFieldsToShow={
                  [
                    "score",
                  ]
                }
                key="20"
                story={
                  {
                    "by": "coolUser",
                    "descendants": 10,
                    "id": 20,
                    "score": 4,
                    "time": 1666689031,
                    "title": "someTitle",
                    "url": "https://reactjs.org/",
                  }
                }
                to="/hacker-news/20"
              >
                <ForwardRef
                  className="story color-grey3 link"
                  loading={false}
                  onClick={[Function]}
                >
                  <div
                    className="ant-card ant-card-bordered story color-grey3 link"
                    onClick={[Function]}
                  >
                    <div
                      className="ant-card-body"
                      style={{}}
                    >
                      <span
                        className="title color-orange bold mb-10"
                      >
                        someTitle
                      </span>
                      <span>
                        4
                         point
                        s
                         | 
                        <span
                          className="user bold"
                        >
                          coolUser
                        </span>
                         | 
                        25.10.2022
                      </span>
                    </div>
                  </div>
                </ForwardRef>
              </StoryCard>
            </div>
          </div>
        </div>
      </Main>
    </Router>
  </BrowserRouter>
</Provider>
`;
